base/linalg/dense.jl

```diff
@@ -693,6 +690,309 @@ function inv(A::StridedMatrix{T}) where T
 end
 
 """
+    cos(A::StridedMatrix)
+
+Compute the matrix cosine of a square matrix `A`.
+
+If `A` is symmetric or Hermitian, its eigendecomposition ([`eigfact`](@ref)) is used to
+compute the cosine. Otherwise, the cosine is determined by calling [`exp`](@ref).
+
+# Examples
+```jldoctest
+julia> cos(ones(2, 2))
+2×2 Array{Float64,2}:
+  0.291927  -0.708073
+ -0.708073   0.291927
+```
+"""
+function cos(A::StridedMatrix{<:Real})
+    if issymmetric(A)
+        return full(cos(Symmetric(A)))
+    end
+    return real(exp!(im*A))
+end
+function cos(A::StridedMatrix{<:Complex})
+    if ishermitian(A)
+        return full(cos(Hermitian(A)))
+    end
+    return (exp!(im*A) + exp!(-im*A)) / 2
+end
+
+"""
+    sin(A::StridedMatrix)
+
+Compute the matrix sine of a square matrix `A`.
+
+If `A` is symmetric or Hermitian, its eigendecomposition ([`eigfact`](@ref)) is used to
+compute the sine. Otherwise, the sine is determined by calling [`exp`](@ref).
+
+# Examples
+```jldoctest
+julia> sin(ones(2, 2))
+2×2 Array{Float64,2}:
+ 0.454649  0.454649
+ 0.454649  0.454649
+```
+"""
+function sin(A::StridedMatrix{<:Real})
+    if issymmetric(A)
+        return full(sin(Symmetric(A)))
+    end
+    return imag(exp!(im*A))
+end
+function sin(A::StridedMatrix{<:Complex})
+    if ishermitian(A)
+        return full(sin(Hermitian(A)))
+    end
+    return (exp!(im*A) - exp!(-im*A)) / 2im
+end
+
+"""
+    sincos(A::StridedMatrix)
+
+Compute the matrix sine and cosine of a square matrix `A`.
+
+# Examples
+```jldoctest
+julia> S, C = sincos(ones(2, 2));
+
+julia> S
+2×2 Array{Float64,2}:
+ 0.454649  0.454649
+ 0.454649  0.454649
+
+julia> C
+2×2 Array{Float64,2}:
+  0.291927  -0.708073
+ -0.708073   0.291927
+```
+"""
+function sincos(A::StridedMatrix{<:Real})
+    if issymmetric(A)
+        return full.(sincos(Symmetric(A)))
+    end
+    c, s = reim(exp!(im*A))
+    return s, c
+end
+function sincos(A::StridedMatrix{<:Complex})
+    if ishermitian(A)
+        return full.(sincos(Hermitian(A)))
+    end
+    X, Y = exp!(im*A), exp!(-im*A)
+    return (X - Y) / 2im, (X + Y) / 2
```

Comment:
Not sure we care, but you allocate four more matrices than necessary here: you could modify `X` and `Y` in-place to be `(X-Y)/2im` and `(X+Y)/2` with a loop:
```jl
@inbounds for i in eachindex(X)
    x, y = X[i], Y[i]
    X[i] = (y-x) * 0.5im
    Y[i] = (y+x) * 0.5
end
return X, Y
```
and similarly elsewhere.

Reply:
Also, wouldn't it be faster to do
```jl
X = exp!(im*A)
Y = inv(X)
```
since surely `inv` is faster than `exp!`?   There might be an accuracy problem if `X` is ill-conditioned, but wouldn't `exp!` have problems too in that case?

(Similarly below.)